saga/saga.go

```diff
@@ -195,54 +197,85 @@ func (s *Saga) updateSagaState(msgs []SagaMessage) error {
 // is one per saga currently executing.  This ensures all updates are applied
 // in order to a saga.  Also controls access to the SagaState so its is
 // updated in a thread safe manner
+// Processes at most SagaUpdateChSize number of updates at a time
 func (s *Saga) updateSagaStateLoop() {
+	updates := []sagaUpdate{}
 	for update := range s.updateCh {
-		s.updateSaga(update)
+		updates = append(updates, update)
+		if len(s.updateCh) == 0 || len(updates) == common.DefaultSagaUpdateChSize {
+			s.updateSaga(updates)
+			updates = []sagaUpdate{}
+		}
 	}
 }
 
-// updateSaga updates the saga s by applying update atomically and sending any error to the requester
-func (s *Saga) updateSaga(update sagaUpdate) {
+// updateSaga updates the saga s by applying updates atomically and sending any error to the requester
+// it checks if the message(s) is a valid transition and batch logs the messages from all updates durably to the SagaLog
+// if msg is an invalid transition, it will neither log nor update internal state
+// always returns the new SagaState that should be used, either the mutated one or a copy of the original.
+func (s *Saga) updateSaga(updates []sagaUpdate) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
+
+	// updateSaga will mutate state if it's a valid transition, but if we then error logging,
+	// we'll need to revert to old state
+	oldState := copySagaState(s.state)
+
 	var err error
-	s.state, err = logMessages(s.state, update.msgs, s.log)
-	update.resultCh <- err
+	var validUpdates []sagaUpdate
+	var validUpdateMsgs []SagaMessage
+
+	for _, update := range updates {
+		err = nil
+
+		// verify that the applied message(s) results in a valid state
+		if len(update.msgs) == 1 {
```

Comment:
Added this separate case to avoid copying sagaState(which takes some time) in case of single message, and call updateSagaState directly. Most updates will contain a single message. The only place I see bulk messages being used is while killing multiple jobs and aborting tasks that is called by getTaskAssignments(). 

If we wanna completely skip the copySagaState() call in bulkUpdateSagaState function, we can avoid this case of len == 1 without any impact on time. I'm up for either way.